### PR TITLE
Merge 5a32892

### PR DESCRIPTION
**Pull request by FraMecca**
This merge includes full compatibility with the precise GC implementation (alongside the conservative).

The code used for profiling was removed (a typeinfo allocation caused a deadlock) and as discussed new code for profiling will be the focus of the next PR.

Even with the simple profiler that is currently available, results are positive:

Voronoi -- Conservative
```
	Number of collections:  8
	Total GC prep time:  2 milliseconds
	Total mark time:  516 milliseconds
	Total sweep time:  198 milliseconds
	Total page recovery time:  106 milliseconds
	Max Pause Time:  298 milliseconds
	Grand total GC time:  824 milliseconds
GC summary:  247 MB,    8 GC  824 ms, Pauses  518 ms <  298 ms
```

Voronoi -- Conservative -- Concurrent
```
Number of collections:  5
	Total GC prep time:  2 milliseconds
	Total mark time:  21 milliseconds
	Total sweep time:  192 milliseconds
	Total page recovery time:  111 milliseconds
	Max Pause Time:  22 milliseconds
	Grand total GC time:  327 milliseconds
GC summary:  287 MB,    5 GC  327 ms, Pauses   23 ms <   22 ms
```

Dustmite -- Conservative
```
Number of collections:  9
	Total GC prep time:  4 milliseconds
	Total mark time:  573 milliseconds
	Total sweep time:  496 milliseconds
	Total page recovery time:  188 milliseconds
	Max Pause Time:  338 milliseconds
	Grand total GC time:  1263 milliseconds
GC summary:  477 MB,    9 GC 1263 ms, Pauses  577 ms <  338 ms
```

Dustmite -- Precise
```
	Number of collections:  9
	Total GC prep time:  4 milliseconds
	Total mark time:  549 milliseconds
	Total sweep time:  456 milliseconds
	Total page recovery time:  162 milliseconds
	Max Pause Time:  310 milliseconds
	Grand total GC time:  1173 milliseconds
GC summary:  477 MB,    9 GC 1173 ms, Pauses  554 ms <  310 ms
```

Dustmite -- Conservative -- Concurrent
```
	Number of collections:  6
	Total GC prep time:  4 milliseconds
	Total mark time:  30 milliseconds
	Total sweep time:  480 milliseconds
	Total page recovery time:  250 milliseconds
	Max Pause Time:  32 milliseconds
	Grand total GC time:  766 milliseconds
GC summary:  477 MB,    6 GC  766 ms, Pauses   35 ms <   32 ms
```

Dustmite -- Precise -- Concurrent
```
	Number of collections:  6
	Total GC prep time:  4 milliseconds
	Total mark time:  33 milliseconds
	Total sweep time:  506 milliseconds
	Total page recovery time:  276 milliseconds
	Max Pause Time:  35 milliseconds
	Grand total GC time:  821 milliseconds
GC summary:  477 MB,    6 GC  821 ms, Pauses   38 ms <   35 ms
```

My three commits are the following:
* [fixed gc.d up to 5a32892](https://github.com/FraMecca/druntime/commit/d8c27862655f6d03e076e30010d69b5e0e69a1e6) where I made all the changes to gc.d in order to get my code up to date with upstream. This is the file that should be reviewed and includes both my code and Rainers' code;
* [removed stats.d file](https://github.com/FraMecca/druntime/commit/660a2e6ac2dc8a815deac7c96cab11cb0142ab99)
* [https://github.com/FraMecca/druntime/commit/1bdc732e86561df3bebfd06c1b3508b0f130726f](https://github.com/FraMecca/druntime/commit/1bdc732e86561df3bebfd06c1b3508b0f130726f) that is the result of git merge with the only addition of having every function in os.d marked as @nogc (as was done by Rainers).

### CODE REVIEW

**Comment by llucax** [comment]
It would be nice to see also what was the total run time for each of the tested programs. Can you include that info too?

**Comment by llucax** [comment]
I also don't get why are you including almost 400 commits. Can't you rebase your commits on top of the current druntime master?

**Comment by FraMecca** [comment]
This is dustmite being timed with time -pv

```
Done in 25 tests and 22 secs and 797 ms; reduced version is in ../../druntime/src.reduced
	Number of collections:  6
	Total GC prep time:  2 milliseconds
	Total mark time:  0 milliseconds
	Total sweep time:  434 milliseconds
	Total page recovery time:  254 milliseconds
	Max Pause Time:  0 milliseconds
	Grand total GC time:  691 milliseconds
GC summary:  532 MB,    6 GC  691 ms, Pauses    2 ms <    0 ms
	Command being timed: "./dustmite --force ../../druntime/src/ grep -nR HAVE_FORK --DRT-gcopt=fork:1 --DRT-gcopt=profile:1"
	User time (seconds): 22.10
	System time (seconds): 1.97
	Percent of CPU this job got: 103%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:23.23
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 510224
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 686933
	Voluntary context switches: 132
	Involuntary context switches: 3682
	Swaps: 0
	File system inputs: 0
	File system outputs: 63984
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

```
Done in 25 tests and 22 secs and 916 ms; reduced version is in ../../druntime/src.reduced
	Number of collections:  10
	Total GC prep time:  3 milliseconds
	Total mark time:  619 milliseconds
	Total sweep time:  549 milliseconds
	Total page recovery time:  192 milliseconds
	Max Pause Time:  316 milliseconds
	Grand total GC time:  1364 milliseconds
GC summary:  532 MB,   10 GC 1364 ms, Pauses  622 ms <  316 ms
	Command being timed: "./dustmite --force ../../druntime/src/ grep -nR HAVE_FORK --DRT-gcopt=fork:0 --DRT-gcopt=profile:1"
	User time (seconds): 21.66
	System time (seconds): 1.54
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:23.26
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 461540
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 461497
	Voluntary context switches: 121
	Involuntary context switches: 2319
	Swaps: 0
	File system inputs: 0
	File system outputs: 63984
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

File: src/gc/impl/conservative/gc.d [hunk 1]
```diff
@@ -2531,10 +2531,10 @@ Lmark:
...
             if (!shouldFork)
             {
-		if (ConservativeGC.isPrecise)
-		    markAll!markPrecise(nostack);
-		else
-		    markAll!markConservative(nostack);
+				if (ConservativeGC.isPrecise)
+					markAll!markPrecise(nostack);
+				else
+					markAll!markConservative(nostack);
```
Comment (llucax) [inline src/gc/impl/conservative/gc.d]: The indentation is actually broken here too, you are using `\t` (tabs) instead of ` ` (spaces).

This applies for the whole commit.

File: src/gc/impl/conservative/gc.d [hunk 2]
```diff
@@ -1761,8 +1761,8 @@ struct Gcx
         if (npages > 1)
             memset(&pool.pagetable[pn + 1], B_PAGEPLUS, npages - 1);
         pool.mark.set(pn);
-        // pool.updateOffsets(pn);
         usedLargePages += npages;
+		pool.freepages -= npages;
```
Comment (llucax) [inline src/gc/impl/conservative/gc.d]: This is not a comment, typo or indentation, so it should probably go in a separate commit.

File: src/gc/bits.d [hunk 3]
```diff
@@ -93,6 +94,280 @@ struct GCBits
         return core.bitop.btr(data, i);
     }
 
+    mixin template RangeVars()
```
Comment (llucax) [inline src/gc/bits.d]: All these additions are missing some documentation.

In particular for all functions ending with `Z`, what does the `Z` means? Also, there are no tests for them, should they be private?

**Review by llucax — COMMENTED** [review]
I did the review as far as I could given the current state of the PR, which include some merges that complicate the review.

File: src/gc/bits.d (hunk 3 above)
Comment (FraMecca) [reply src/gc/bits.d]: This was done by Rainers.
